pallene/function_inline.lua

```diff
@@ -0,0 +1,179 @@
+    return ir.map_cmd(called,function (_cmd)
+        if _cmd._tag == 'ir.Cmd.Return' then
+            local moves = {}
+            for k,v in ipairs(_cmd.srcs) do
+                -- if the function call will put the results into a,b,c
+                -- and the called returns x,y,z
+                -- this creates moves corresponding to x = a, y = b and z = c
+                table.insert(moves,ir.Cmd.Move(_cmd.loc,cmd.dsts[k],v))
+            end
+            return ir.Cmd.Seq(moves)
+        end
+    end)
+end
+
+
+-- find_top finds the highest variable index in the 'cmd' subtree
+local function find_top(cmd)
+    local top = 0
+    local function change_top(val)
+        if val > top then top = val end
+    end
+    local function findtop_localvar(val)
+        if type(val) == "table" and val._tag == 'ir.Value.LocalVar' then
+            change_top(val.id)
+        end
+    end
+    local function findtop(_cmd)
+        if _cmd._tag == 'ir.Cmd.Binop' then
+            change_top(_cmd.dst)
+            findtop_localvar(_cmd.src1)
+            findtop_localvar(_cmd.src2)
+            return _cmd
+        end
+        if _cmd._tag == 'ir.Cmd.Move' then
+            change_top(_cmd.dst)
+            return _cmd
+        end
+        if _cmd._tag == 'ir.Cmd.Return' then
+
+            for _, value in pairs(_cmd.srcs) do
+                findtop_localvar(value)
+            end
+            return _cmd
+        end
+    end
+    ir.map_cmd(cmd,findtop)
+    return top
+end
+
+-- auxiliary function to transform values
+local function map_val(cmd,fval,flocalval)
+    return ir.map_cmd(cmd,function (_cmd)
+        if _cmd._tag == 'ir.Cmd.Binop' then
+            _cmd.src1 = flocalval(_cmd.src1)
+            _cmd.src2 = flocalval(_cmd.src2)
+            _cmd.dst = fval(_cmd.dst)
+            return _cmd
+        end
+        if _cmd._tag == 'ir.Cmd.Move' then
+            _cmd.src = flocalval(_cmd.src)
+            _cmd.dst = fval(_cmd.dst)
+            return _cmd
+        end
+        if _cmd._tag == 'ir.Cmd.Return' then
+            for key, value in pairs(_cmd.srcs) do
+                _cmd.srcs[key] = flocalval(value)
+            end
+            return _cmd
+        end
+    end)
+end
+
+-- adjusts the function body to be inlined in respect to the outer function top
+-- so the internal variables will not collide with the outer ones
+local function adjust(cmd,top)
+    local new_val = function(val)
+        return top+val
+    end
+    local iflocalvar_new_val = function (val)
+        if type(val) == "table" and val._tag == 'ir.Value.LocalVar' then
+            val.id = new_val(val.id)
+        end
+        return val
+    end
+    return map_val(cmd,new_val,iflocalvar_new_val)
+end
+
+-- normal_form guarantess that the cmd subtree will be in normal form
+-- starting at 1 and every value is increased in order of appearance, without holes
+local function normal_form(cmd)
+    local ids = {}
+    ids.top = 1
```

Comment:
I think it would be better to put the next index in a separate variable (`local top = 1`) instead of also putting it inside the `ids` table.

Reply:
when I change `ids.top = 1` to `local top = 1` it isn't read as an upvalue (it is nil). Am I missing something?

Reply:
Hard for me to say without seeing the code. But it should work.